src/components/MultiHopTrade/hooks/useReceiveAddress.tsx

```diff
@@ -32,65 +33,60 @@ export const useReceiveAddress = ({
   buyAccountId: AccountId | undefined
   buyAsset: Asset | undefined
 }) => {
-  const wallet = useWallet().state.wallet
-  const [walletReceiveAddress, setWalletReceiveAddress] = useState<string | undefined>(undefined)
+  const { wallet, deviceId } = useWallet().state
   const buyAccountMetadataFilter = useMemo(() => ({ accountId: buyAccountId }), [buyAccountId])
   const buyAccountMetadata = useAppSelector(state =>
     selectPortfolioAccountMetadataByAccountId(state, buyAccountMetadataFilter),
   )
 
-  const getReceiveAddressFromBuyAsset = useCallback(
-    async (buyAsset: Asset) => {
-      if (!wallet) {
-        return
-      }
-      if (!buyAccountId) {
-        return
-      }
-      if (!buyAccountMetadata) {
-        return
-      }
-      if (isUtxoAccountId(buyAccountId) && !buyAccountMetadata.accountType)
-        throw new Error(`Missing accountType for UTXO account ${buyAccountId}`)
-      const buyAssetChainId = buyAsset.chainId
-      const buyAssetAccountChainId = fromAccountId(buyAccountId).chainId
-      /**
-       * do NOT remove
-       * super dangerous - don't use the wrong bip44 params to generate receive addresses
-       */
-      if (buyAssetChainId !== buyAssetAccountChainId) {
-        return
-      }
+  // This flag is used to skip the query below and treat missing input as `isLoading` to prevent UI
+  // flashing during state changes. Any of the conditions below returning true should be treated as
+  // "we're not yet ready to determine if a wallet receive address is available"
+  const isInitializing = useMemo(() => {
+    if (!buyAsset || !wallet || !buyAccountId || !buyAccountMetadata) {
+      return true
+    }
 
-      const fetchUnchainedAddress = Boolean(wallet && isLedger(wallet))
+    const buyAssetChainId = buyAsset.chainId
+    const buyAssetAccountChainId = fromAccountId(buyAccountId).chainId
 
-      const receiveAddress = await getReceiveAddress({
-        asset: buyAsset,
-        wallet,
-        accountMetadata: buyAccountMetadata,
-        deviceId: await wallet.getDeviceID(),
-        pubKey: fetchUnchainedAddress ? fromAccountId(buyAccountId).account : undefined,
-      })
-      return receiveAddress
-    },
-    [buyAccountId, buyAccountMetadata, wallet],
-  )
+    /**
+     * do NOT remove
+     * super dangerous - don't use the wrong bip44 params to generate receive addresses
+     */
+    if (buyAssetChainId !== buyAssetAccountChainId) {
+      return true
+    }
+
+    return false
+  }, [buyAccountId, buyAccountMetadata, buyAsset, wallet])
+
+  const { data: walletReceiveAddress, isLoading } = useQuery({
+    queryKey: ['receiveAddress', buyAsset?.assetId, deviceId, buyAccountId],
+    queryFn: isInitializing
+      ? skipToken
+      : async () => {
+          // Already covered in shouldSkip, but TypeScript lyfe mang.
```

Comment:
Is this comment still valid? There is no `shouldSkip` reference

Reply:
Also if we want type safety here we can explicitly re-add the checks from `isInitializing` (i.e the checks below` in addition to `isInitializing` itself i.e

```suggestion
`isInitializing && (!buyAsset || !wallet || !buyAccountId || !buyAccountMetadata)`
```